
1. Load libraries:

```{r, message = FALSE}
require(ggplot2)
```

2. Load data:

```{r, message = FALSE}
data <- read.table(file = "data/gene_lists/ensembl_bter_amel_comp.txt",
                   header = FALSE)
colnames(data) <- c("bter_locus",
                    "orthology_confidence",
                    "whole-genome_alignment_coverage",
                    "Gene-order_conservation_score",
                    "percent_id_target_to_query",
                    "percent_id_query_to_target",
                    "last_common_ancestor",
                    "amel_locus")

## Subset based on orthology confidence:
data <- subset(x = data,
               orthology_confidence == 1)
```

3. Load them against each other:

```{r, message = FALSE}
ggplot(data = data,
       aes(x = percent_id_target_to_query,
           y = percent_id_query_to_target)) +
  geom_point()
```

4. Read in genes of interest:

```{r, message = FALSE}
lowdiv_genes <- scan(file = "data/gene_lists/low_diversity_genes.txt",
                     as.character())

busted_output <- read.table(file = "~/Documents/bt_busted_conserved_selection.tsv",
                            header = TRUE)

busted_subset <- subset(x = busted_output,
                        pval < 0.05)

loss_of_function_genes <- scan(file = "data/gene_lists/loss_of_function_genes.txt",
                     as.character())

bter_amel_plot <- ggplot(data = data,
       aes(x = percent_id_target_to_query,
           y = percent_id_query_to_target)) +
  xlab(label = expression(paste(italic("Apis mellifera"),
                                " sequence similarity (%)"))) +
  ylab(label = expression(paste(italic("Bombus terrestris"),
                                " sequence similarity (%)"))) +
  xlim(c(25,100)) +
  ylim(c(25,100)) +
  geom_point(alpha = 0.7,
             size = 2,
             fill = "grey",
             pch = 21) +
  geom_point(data = subset(x = data,
                           bter_locus %in% lowdiv_genes),
             colour = "green",
             fill = "red",
             size = 2,
             pch = 21,
             alpha = 0.7) +
  geom_point(data = subset(x = data,
                           bter_locus %in% busted_subset$geneID),
             colour = "black",
             fill = "red",
             size = 2,
             pch = 21,
             alpha = 0.7) +
  geom_point(data = subset(x = data,
                           bter_locus %in% lowdiv_genes),
            colour = "black",
             fill = "blue",
             size = 2,
             pch = 21,) +
  geom_point(data = subset(x = data,
                           bter_locus %in% loss_of_function_genes),
             colour = "black",
             fill = "yellow",
             size = 2,
             pch = 21) +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 12)) + 
  stat_cor(p.accuracy = 0.0001)
```

5. Read in data for Drosophila melanogaster:

```{r, message = FALSE}
dmel_data <- read.table(file = "data/gene_lists/ensembl_bter_dmel_comp.txt",
                   header = FALSE)
colnames(dmel_data) <- c("bter_locus",
                    "orthology_confidence",
                    "percent_id_target_to_query",
                    "percent_id_query_to_target",
                    "last_common_ancestor",
                    "dmel_locus")

## Subset based on orthology confidence:
dmel_data <- subset(x = dmel_data,
               orthology_confidence == 1)
```

3. Load them against each other:

```{r, message = FALSE}
ggplot(data = dmel_data,
       aes(x = percent_id_target_to_query,
           y = percent_id_query_to_target)) +
  geom_point()
```

4. Read in genes of interest:

```{r, message = FALSE}
bter_dmel_plot <- ggplot(data = dmel_data,
       aes(x = percent_id_target_to_query,
           y = percent_id_query_to_target)) +
  xlab(label = expression(paste(italic("Drosophila melanogaster"),
                                " sequence similarity (%)"))) +
  ylab(label = expression(paste(italic("Bombus terrestris"),
                                " sequence similarity (%)"))) +
  xlim(c(25,100)) +
  ylim(c(25,100)) +
  geom_point(alpha = 0.7,
             size = 2,
             fill = "grey",
             pch = 21) +
  geom_point(data = subset(x = dmel_data,
                           bter_locus %in% lowdiv_genes),
             colour = "black",
             fill = "green",
             size = 2,
             pch = 21,
             alpha = 0.7) +
  geom_point(data = subset(x = dmel_data,
                           bter_locus %in% busted_subset$geneID),
             colour = "black",
             fill = "red",
             size = 2,
             pch = 21,
             alpha = 0.7) +
  geom_point(data = subset(x = dmel_data,
                           bter_locus %in% lowdiv_genes),
            colour = "black",
             fill = "blue",
             size = 2,
             pch = 21,) +
  geom_point(data = subset(x = dmel_data,
                           bter_locus %in% loss_of_function_genes),
             colour = "yellow",
             shape = 22) +
  theme_bw() + 
  stat_cor(p.accuracy = 0.0001)

ggarrange(bter_amel_plot,
          bter_dmel_plot,
          ncol = 2,
          nrow = 1,
          labels = c("A",
                     "B"),
          align = "hv")
```

Compare homology across genes:

```{r, message = FALSE}
bee_comp <- read.table(file = "data/loss_of_function_genes_comp.txt",
                       header = TRUE)
bee_comp$HOG <- NULL
bee_comp$Gene_Tree_Parent_Clade <- NULL

library(reshape2)

tmp <- bee_comp[c(3:5), ]
tmp_melt <- melt(tmp, id.vars = "OG")
tmp_melt$value <- as.numeric(tmp_melt$value)

tmp_melt$variable <- gsub(pattern = "_",
                       replacement = " ",
                       tmp_melt$variable)

## Reorder species:
tmp_melt$variable <- factor(tmp_melt$variable,
                            levels = c("Apis mellifera",
                                       "Bombus terrestris",
                                       "Ceratina calcarata",
                                       "Eufriesea mexicana",
                                       "Habropoda laboriosa",
                                       "Megachile rotundata",
                                       "Osmia bicornis",
                                       "Osmia lignaria",
                                       "Dufourea novaeangliae",
                                       "Megalopta genalis",
                                       "Nomia melanderi"))

heatmap <- ggplot(tmp_melt,
       aes(x = OG,
           y = variable,
           fill = value)) +
  geom_tile(color = "black") +
  geom_text(aes(label = value),
            color = "white",
            size = 4) +
  coord_fixed() +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.y = element_text(face = "italic"),
        axis.text.x = element_text(angle = 90),
        axis.title = element_blank()) +
  scale_y_discrete(limits=rev)

ggarrange(bter_amel_plot,
          heatmap,
          ncol = 2,
          nrow = 1,
          widths = c(1.5,1),
          labels = c("A",
                     "B"))

ggsave(filename = "comp_test.pdf",
       height = 4,
       width = 6)

```