---
author: "Joe Colgan"
title: "**Visualisation of output from nSL-based analysis**"
output:
  pdf_document: default
  html_document: default
fig_width: 4
fig_height: 4
fontsize: 20pt
---

## Introduction
The purpose of the present analysis is to analyse and visualise the output of the nSL-based analysis for canonical
immune genes found in the genome of the earth/buff-tailed bumblebee _Bombus terrestris_.  

1. Load libraries:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
library(ggplot2)
library(ggpubr)
library(viridis)
library(dplyr)
```

2. Load data:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Set relative path to input file:
input <- "data/selection_analyses/nsl_score_abs_top_score_per_gene.txt"

## Read in input data:
data <- read.table(file = input,
                   header = FALSE)

## Assign column names:
colnames(data) <- c("chromosome",
                    "position",
                    "variant_type",
                    "modifier",
                    "locus",
                    "nsl_score",
                    "significance")

## Read in file containing all nsl-annotated SNPs:
snp_ann <- "data/selection_analyses/nsl_score_abs_with_annotations.txt"
all_snp_data <- read.table(file = snp_ann,
                           header = FALSE)
## Update header information:
colnames(all_snp_data) <- c("chromosome",
                            "position",
                            "variant_type",
                            "modifier",
                            "locus",
                            "nsl_score",
                            "significance")
```

3. Load immune genes:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Read in a list of canonical immune genes and associated gene groups: 
immune_genes <- scan(file = "data/gene_lists/barribeau_2015_gene_symbols.txt",
                     as.character())

## Add missing IDs for those genes not annotated with a gene symbol 
## starting with 'LOC':
immune_genes <- c(immune_genes,
                  "Def",
                  "VSP")
```

4. Read in population genetic measures, including pi and Tajima's D:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Set relative path to input file:
pop_data <- read.table(file = "data/selection_analyses/pi_estimates_per_gene_reformatted_subsetted_with_genes.txt",
                       header = FALSE)
## Remove redundant fields:
pop_data$V5 <- NULL
pop_data$V6 <- NULL
pop_data$V7 <- NULL

## Update column names:
colnames(pop_data) <- c("chromosome",
                        "start_position",
                        "end_position",
                        "locus",
                        "pi",
                        "tajimas_d",
                        "gene_length",
                        "length_corrected_theta")

## Correct theta estimates for gene length:
pop_data$length_corrected_pi <- pop_data$pi/pop_data$gene_length
```

5. Match genes:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Add top |nsl| score per gene to the dataframe containing pop genetic output: 
pop_data$nsl <- data[match(pop_data$locus,
                           data$locus), ]$nsl_score

## Plot the relationship between Tajima's D and nSL score across all genes:
ggplot(data = pop_data,
       aes(x = tajimas_d,
           y = nsl)) +
  xlab(label = "Tajima's D") +
  ylab(label = "|nsl| score") +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "gam") +
  theme_bw()

## Subset immune genes:
immune_genes_df <- subset(x = pop_data,
                          locus %in% immune_genes)
## Check structure:
str(immune_genes_df)

## Classify as 'immune':
immune_genes_df$status <- "immune"

## Subset non-immune genes:
non_immune_genes_df <- subset(x = pop_data,
                          !(locus %in% immune_genes))
## Classify as 'non-immune':
non_immune_genes_df$status <- "non-immune"

## Combine:
combined_immune_df <- rbind(immune_genes_df,
                            non_immune_genes_df)

## Turn off scientific notation for plots:
options(scipen = 999)

## Examine differences in |nsl| scores between immune and non-immune genes:
wilcox.test(immune_genes_df$nsl,
            non_immune_genes_df$nsl)

## Visualise using box-plots and/or violin plots:
## Plot status against absolute nSL score:
nsl_comp_plot  <- ggplot(data = combined_immune_df,
                         aes(x = status,
                             y = nsl,
                             colour = status)) +
                         geom_violin() +
                         geom_boxplot(width = 0.1,
                                      color = "grey",
                                      alpha = 0.2) +
  xlab(label = "Gene category") +
  ylab(label = "|nSL| score") +
    scale_fill_viridis(discrete = TRUE) +
  scale_colour_manual(values = c("blue",
                                 "orange")) +
  stat_compare_means() +
  theme_bw() +
  theme(axis.text = element_text(face = "plain",
                                 size = 12),
        axis.title = element_text(face = "bold",
                                  size = 12),
        legend.position = "none")

## Plot status against length corrected nucleotide diversity:
pi_comp_plot <- ggplot(data = combined_immune_df,
       aes(x = status,
           y = length_corrected_pi,
           colour = status)) +
           geom_violin() +
  geom_boxplot(width = 0.1,
               color = "grey",
               alpha = 0.2) +
  xlab(label = "Gene category") +
  ylab(label = "Nucleotide diversity") +
    scale_fill_viridis(discrete = TRUE) +
  scale_colour_manual(values = c("blue",
                                 "orange")) +
  stat_compare_means() +
  theme_bw() +
  theme(axis.text = element_text(face = "plain",
                                 size = 12),
        axis.title = element_text(face = "bold",
                                  size = 12),
        legend.position = "none")

## Plot status against Tajima's D:
taj_comp_plot <- ggplot(data = combined_immune_df,
       aes(x = status,
           y = tajimas_d,
           colour = status)) +
           geom_violin() +
  geom_boxplot(width = 0.1,
               color = "grey",
               alpha = 0.2) +
  xlab(label = "Gene category") +
  ylab(label = "Tajima's D") +
    scale_fill_viridis(discrete = TRUE) +
  scale_colour_manual(values = c("blue",
                                 "orange")) +
  stat_compare_means() +
  theme_bw() +
  theme(axis.text = element_text(face = "plain",
                                 size = 12),
        axis.title = element_text(face = "bold",
                                  size = 12),
        legend.position = "none")

## Plot status against gene length:
length_comparison_plot <- ggplot(data = combined_immune_df,
       aes(x = status,
           y = gene_length,
           colour = status)) +
           geom_violin() +
  geom_boxplot(width = 0.1,
               color = "grey",
               alpha = 0.2) +
  xlab(label = "Gene category") +
  ylab(label = "Gene length (bp)") +
    scale_fill_viridis(discrete = TRUE) +
  scale_colour_manual(values = c("blue",
                                 "orange")) +
  stat_compare_means() +
  scale_y_continuous(trans = "log10") +
  theme_bw() +
  theme(axis.text = element_text(face = "plain",
                                 size = 12),
        axis.title = element_text(face = "bold",
                                  size = 12))

## Check length distribution for immune and non-immune genes:
ggplot(data = combined_immune_df,
       aes(x = gene_length,
           colour = status)) +
           geom_density() +
  xlab(label = "Gene length (bp)") +
  theme_bw() +
  theme(axis.title = element_text(face = "bold",
                                  size = 12),
        legend.position = "top") +
  geom_vline(xintercept = median(subset(x = combined_immune_df,
               status == "immune")$gene_length),
             colour = "blue",
             linetype = "dashed") +
  geom_vline(xintercept = median(subset(x = combined_immune_df,
               status == "non-immune")$gene_length),
             colour = "orange",
             linetype = "dashed") +
  scale_x_continuous(trans = 'log10') +
  scale_colour_manual(values = c("blue", "orange"))

## Pick 162 random non-immune genes of similar length:
summary(subset(x = combined_immune_df,
               status == "immune")$gene_length)
summary(subset(x = combined_immune_df,
               status == "non-immune")$gene_length)

## Find genes of similar length:
combined_size_subset <- subset(x = combined_immune_df,
                               gene_length >= min(subset(combined_immune_df,
                                                         status == "immune")$gene_length) &
                               gene_length <= max(subset(combined_immune_df,
                                                         status == "immune")$gene_length))
summary(combined_size_subset$gene_length)
nrow(combined_size_subset)

tmp <- data.frame()
for (name in subset(x = combined_immune_df,
               status == "immune")$locus){
                 print(name)
    gene_length <- subset(x = combined_immune_df,
                          locus == name)$gene_length
    print(gene_length)
    max_gene_length <- gene_length + 100
    print(max_gene_length)
    min_gene_length <- gene_length - 100
    print(min_gene_length)
    tmp_2 <- subset(x = combined_immune_df,
               status == "non-immune" &
             gene_length < max_gene_length &
             gene_length > min_gene_length)
    print(nrow(tmp_2))
    if (nrow(tmp_2) == 0){
      print(paste(name, " has no match!"))
            } else {
    tmp <- rbind(tmp,
                 sample_n(tmp_2, 1))
    print(tmp)
            }
}

## Combine length-matched non-immune genes with immune genes:
new_combined_df <- rbind(tmp,
                         subset(x = combined_immune_df,
                                status == "immune"))

## Plot status against nucleotide diversity for size-matched genes:
ggplot(data = new_combined_df,
       aes(x = status,
           y = length_corrected_theta,
           colour = status)) +
           geom_violin() +
  geom_boxplot(width = 0.1,
               color = "grey",
               alpha = 0.2) +
  xlab(label = "Gene category") +
  ylab(label = "Length-corrected nucleotide diversity") +
    scale_fill_viridis(discrete = TRUE) +
  scale_colour_manual(values = c("blue",
                                 "orange")) +
  stat_compare_means() +
  scale_y_continuous(trans='log10') +
  theme_bw() +
  theme(axis.text = element_text(face = "plain",
                                 size = 12),
        axis.title = element_text(face = "bold",
                                  size = 12),
        legend.position = "none")

## Plot status against Tajima's D for size-matched genes:
ggplot(data = new_combined_df,
       aes(x = status,
           y = tajimas_d,
           colour = status)) +
           geom_violin() +
  geom_boxplot(width = 0.1,
               color = "grey",
               alpha = 0.2) +
  xlab(label = "Gene category") +
  ylab(label = "Tajima's D") +
    scale_fill_viridis(discrete = TRUE) +
  scale_colour_manual(values = c("blue",
                                 "orange")) +
  stat_compare_means() +
  scale_y_continuous(trans='log10') +
  theme_bw() +
  theme(axis.text = element_text(face = "plain",
                                 size = 12),
        axis.title = element_text(face = "bold",
                                  size = 12),
        legend.position = "none")

## Check if significant difference in nucleotide diversity between:
wilcox.test(immune_genes_df$length_corrected_pi,
            non_immune_genes_df$length_corrected_pi)

wilcox.test(subset(x = new_combined_df,
                   status == "immune")$length_corrected_pi,
            subset(x = new_combined_df,
                   status == "non-immune")$length_corrected_pi)
```

6. Examine how many canonical immune genes are in top 1%, 5% and 10% across all genome-wide SNPs:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Identify the nsl threshold for top 10% SNPs cut-off:
nsl_quantile <- quantile(na.omit(pop_data$nsl),
                         0.90)

## Subset immune genes based on threshold:
immune_genes_top_ten_percent <- subset(x = immune_genes_df,
                                       nsl > nsl_quantile)
## Calculate the number of rows:
dim(immune_genes_top_ten_percent)

## Subset non-immune genes based on threshold:
non_immune_genes_top_ten_percent <- subset(x = non_immune_genes_df,
                                       nsl > nsl_quantile)
## Calculate the dimensions of the resulting dataframe:
dim(non_immune_genes_top_ten_percent)
```

7. Examine nSL values across gene groups:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Read in list of gene groups as identified by Sadd et al. (2015):
gene_groups <- read.table(file = "data/gene_lists/barribeau_groups_added.txt",
                          header = FALSE)

## Update column names:
colnames(gene_groups) <- c("gene_groups",
                           "locus")

## Match two dataframes based on locus name:
tmp <- pop_data[match(gene_groups$locus,
                      pop_data$locus), ]
                      
## Add a 'gene_group' column:
tmp$gene_group <- gene_groups$gene_groups

## Remove 'NA' entries:
tmp <- na.omit(tmp)

## Check temporary dataframe:
head(tmp)

## Plot data:
nsl_scatterplot <- ggplot(data = tmp,
       aes(x = gene_group,
           y = nsl)) +
  geom_point(alpha = 0.5) +
  labs(x = "Gene group") +
  labs(y = expression(bold("|")~bolditalic(n)*bold(S[L])*bold("| score"))) +
  ylim(c(0, 4.5)) +
  geom_point(data = subset(x = tmp,
                           nsl >= 2),
             colour = "blue",
             alpha = 0.5) +
  geom_point(data = subset(x = tmp,
                           nsl > 2.51),
             colour = "red",
             alpha = 0.75) +
  geom_point(data = subset(x = tmp,
                           nsl >= quantile(na.omit(all_snp_data$nsl),
                                   0.99)),
             colour = "red",
             alpha = 0.5) +
  geom_hline(yintercept = quantile(na.omit(all_snp_data$nsl),
                                   0.99),
             linetype = "dashed",
             colour = "red") +
  geom_hline(yintercept = 2,
             linetype = "dashed",
             colour = "blue") +
  theme_bw() +
  theme(axis.text.x = element_text(face = "bold",
                                   angle = 90),
        axis.title = element_text(face = "bold",
                                  size = 12))

# In terms of SCP, do we find more with signatures of selection than expected by random chance?
total_scp <- nrow(subset(x = tmp,
                         gene_group == "SCR"))
total_non_scp <- nrow(data) - total_scp

## Subset SCP genes with scores in the top 1% of all genes:
high_scp <- nrow(subset(x = tmp,
                         gene_group == "SCR" &
                          nsl >= 2.51))

high_non_scp <- nrow(subset(x = pop_data,
                          nsl >= 2.51)) - high_scp

## Construct contigency table:
contingency_scp <- rbind(c(high_scp, high_non_scp),
                         c(total_scp - high_scp,
                           total_non_scp - high_non_scp))

## Run chi-squared test:
contingency_scp_chisq <- chisq.test(contingency_scp)
## Print observed values:
contingency_scp_chisq$observed
## Print expected values:
contingency_scp_chisq$expected
## Print results of test:
contingency_scp_chisq
```

8. Next question - do immune genes have more SNPs under selection compared to other genes?

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Check contents of dataframe:
str(all_snp_data)

## Make locus a character:
all_snp_data$locus <- as.character(unlist(all_snp_data$locus))

## Subset immune genes:
immune_snp_data <- subset(x = all_snp_data,
                       locus %in% immune_genes)

## Create an empty list:
count <- list()
for (name in unique(immune_snp_data$locus)){
  print(name)
  count[[name]] <- print(nrow(subset(x = immune_snp_data,
                  locus == name &
                  significance == 1)))
}

## Convert count to a dataframe:
count_df <- as.data.frame(unlist(count))
## Assign rownames, which contain gene symbols, to a new column 'locus':
count_df$locus <- row.names(count_df)
## Order count dataframe based on 'locus':
count_df <- count_df[order(count_df$locus), ]

snp_counts_immune_df <- as.data.frame(table(immune_snp_data$locus))

snp_counts_immune_df$special <- count_df$`unlist(count)`

## For each gene, calculate the proportion of genes with highest |nSL| scores:
snp_counts_immune_df$prop <- snp_counts_immune_df$special /
                             snp_counts_immune_df$Freq
```

9. Perform the same for all genes:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Subset immune genes:
snp_non_immune_df <- subset(x = all_snp_data,
                            !(locus %in% immune_genes))

## Remove intergenic and up/downstream SNPs:
snp_non_immune_df <- subset(snp_non_immune_df,
                            !(variant_type %in% c("intergenic_region")))

count_2 <- list()
for (name in unique(snp_non_immune_df$locus)){
  print(name)
  count_2[[name]] <- print(nrow(subset(x = snp_non_immune_df,
                                       locus == name &
                                         significance == 1)))
}

count_2_df <- as.data.frame(unlist(count_2))
count_2_df$locus <- row.names(count_2_df)
count_2_df <- count_2_df[order(count_2_df$locus), ]
snp_counts_non_immune_df <- as.data.frame(table(snp_non_immune_df$locus))

snp_counts_non_immune_df$special <- count_2_df$`unlist(count_2)`

snp_counts_non_immune_df$prop <- snp_counts_non_immune_df$special /
                                 snp_counts_non_immune_df$Freq

## Examine the data:
ggplot(data = subset(snp_counts_non_immune_df,
                     Freq > 10),
       aes(x = Freq,
           y = prop)) +
  geom_point() +
  geom_point(data = snp_counts_immune_df,
             colour = "red")
```

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Generate plot:
snp_immune_df$annotation_simplified <- immune_snp_data$variant_type

## Update levels for plotting:
levels(snp_immune_df$annotation_simplified)[grep(x = levels(snp_immune_df$annotation_simplified),
                                        pattern = "UTR")] <- "UTR"
levels(snp_immune_df$annotation_simplified)[grep(x = levels(snp_immune_df$annotation_simplified),
                                        pattern = "stream")] <- "5kb up/down-stream"
levels(snp_immune_df$annotation_simplified)[grep(x = levels(snp_immune_df$annotation_simplified),
                                        pattern = "intron")] <- "Intron"
levels(snp_immune_df$annotation_simplified)[grep(x = levels(snp_immune_df$annotation_simplified),
                                        pattern = "intergenic")] <- "Intergenic"
levels(snp_immune_df$annotation_simplified)[grep(x = levels(snp_immune_df$annotation_simplified),
                                        pattern = "non_coding_transcript")] <- "Non-coding RNA"
levels(snp_immune_df$annotation_simplified)[!(levels(snp_immune_df$annotation_simplified) %in% c("UTR",
                                                                               "5kb up/down-stream",
                                                                               "Intron",
                                                                               "Intergenic",
                                                                               "Non-coding RNA"))] <- "Coding"
                                                                               
## Ensure column is a factor and set levels:
snp_immune_df$annotation_simplified <- factor(snp_immune_df$annotation_simplified,
                                     levels = rev(c("Non-coding RNA",
                                                    "Coding",
                                                    "UTR",
                                                    "Intergenic",
                                                    "5kb up/down-stream",
                                                    "Intron")))
                                                    
## Print levels to console:
levels(snp_immune_df$annotation_simplified)
## Convert gene counts to a dataframe:
snp_immune_table <- as.data.frame(table(immune_snp_data$locus))

## Add gene group information:
snp_immune_table$gene_group <- gene_groups[match(snp_immune_table$Var1,
                                                 gene_groups$locus), ]$gene_group

## All non-immune genes:
snp_non_immune_df$annotation_simplified <- snp_non_immune_df$variant_type

## Update levels for plotting:
levels(snp_non_immune_df$annotation_simplified)[grep(x = levels(snp_non_immune_df$annotation_simplified),
                                        pattern = "UTR")] <- "UTR"
levels(snp_non_immune_df$annotation_simplified)[grep(x = levels(snp_non_immune_df$annotation_simplified),
                                        pattern = "stream")] <- "5kb up/down-stream"
levels(snp_non_immune_df$annotation_simplified)[grep(x = levels(snp_non_immune_df$annotation_simplified),
                                        pattern = "intron")] <- "Intron"
levels(snp_non_immune_df$annotation_simplified)[grep(x = levels(snp_non_immune_df$annotation_simplified),
                                        pattern = "intergenic")] <- "Intergenic"
levels(snp_non_immune_df$annotation_simplified)[grep(x = levels(snp_non_immune_df$annotation_simplified),
                                        pattern = "non_coding_transcript")] <- "Non-coding RNA"
levels(snp_non_immune_df$annotation_simplified)[!(levels(snp_non_immune_df$annotation_simplified) %in% c("UTR",
                                                                               "5kb up/down-stream",
                                                                               "Intron",
                                                                               "Intergenic",
                                                                               "Non-coding RNA"))] <- "Coding"
## Ensure column is a factor and set levels:
snp_non_immune_df$annotation_simplified <- factor(snp_non_immune_df$annotation_simplified,
                                     levels = rev(c("Non-coding RNA",
                                                    "Coding",
                                                    "UTR",
                                                    "Intergenic",
                                                    "5kb up/down-stream",
                                                    "Intron")))

## Count the frequencies of different variant types for immune genes:
table(snp_immune_df$annotation_simplified)
## Calculate the sum:
total_all_immune <- sum(table(snp_immune_df$annotation_simplified))
total_all_immune

## Count the frequencies of different variant types for non-immune genes:
table(snp_non_immune_df$annotation_simplified)
## Calculate the sum:
total_all_nonimmune <- sum(table(snp_non_immune_df$annotation_simplified))
total_all_nonimmune

immune_intron_all     <- table(snp_immune_df$annotation_simplified)["Intron"]
non_immune_intron_all <- table(snp_non_immune_df$annotation_simplified)["Intron"]

## Construct a contingency table:
intron_all <- rbind(c(immune_intron_all,
                  non_immune_intron_all),
                c(total_all_immune - immune_intron_all,
                  total_all_nonimmune - non_immune_intron_all))
## Perform test:
intron_all_chi <- chisq.test(intron_all)
## Print result:
intron_all_chi
## Print observed values:
intron_all_chi$observed
## Print expected values:
intron_all_chi$expected

## Perform the same for SNP 5kb up/downstream of SNPs: 
immune_stream_all  <- table(snp_immune_df$annotation_simplified)["5kb up/down-stream"]
non_immune_stream_all  <- table(snp_non_immune_df$annotation_simplified)["5kb up/down-stream"]

## Construct a contingency table:
stream_all <- rbind(c(immune_stream_all,
                  non_immune_stream_all),
                c(total_all_immune - immune_stream_all,
                  total_all_nonimmune - non_immune_stream_all))
## Perform test:
stream_all_chi <- chisq.test(stream_all)
## Print result:
stream_all_chi
## Print observed values:
stream_all_chi$observed
## Print expected values:
stream_all_chi$expected

## Perform the same for SNPs in UTRs: 
immune_utr_all  <- table(snp_immune_df$annotation_simplified)["UTR"]
non_immune_utr_all  <- table(snp_non_immune_df$annotation_simplified)["UTR"]

## Construct a contingency table:
utr_all <- rbind(c(immune_utr_all,
                  non_immune_utr_all),
                c(total_all_immune - immune_utr_all,
                  total_all_nonimmune - non_immune_utr_all))
## Perform test:
utr_all_chi <- chisq.test(utr_all)
## Print result:
utr_all_chi
## Print observed values:
utr_all_chi$observed
## Print expected values:
utr_all_chi$expected

## Perform the same for SNPs in coding regions: 
immune_cds_all  <- table(snp_immune_df$annotation_simplified)["Coding"]
non_immune_cds_all  <- table(snp_non_immune_df$annotation_simplified)["Coding"]

## Construct a contingency table:
cds_all <- rbind(c(immune_cds_all,
                   non_immune_cds_all),
                   c(total_all_immune - immune_cds_all,
                     total_all_nonimmune - non_immune_cds_all))
## Perform test:
cds_all_chi <- chisq.test(cds_all)
## Print result:
cds_all_chi
## Print observed values:
cds_all_chi$observed
## Print expected values:
cds_all_chi$expected
```

If we look at SNPs with signatures of selection:

```{r, message = FALSE}
colnames(snp_immune_df) <- c("chromosome",
                            "position",
                            "variant_type",
                            "modifier",
                            "locus",
                            "nsl_score",
                            "significance",
                            "annotation_simplified")

## Subset SNPs with nSL scores >= 2 for both immune and non-immune genes:
snp_immune_selection_df <- subset(x = snp_immune_df,
                                  significance == 1)
snp_non_immune_selection_df <- subset(x = snp_non_immune_df,
                                      significance == 1)

## Count the frequencies of different variant types for immune genes:
table(snp_immune_selection_df$annotation_simplified)
## Calculate the sum:
total_selected_immune <- sum(table(snp_immune_selection_df$annotation_simplified))
total_selected_immune

## Count the frequencies of different variant types for non-immune genes:
table(snp_non_immune_selection_df$annotation_simplified)
## Calculate the sum:
total_selected_nonimmune <- sum(table(snp_non_immune_selection_df$annotation_simplified))
total_selected_nonimmune

immune_intron_selection  <- table(snp_immune_selection_df$annotation_simplified)["Intron"]
non_immune_intron_selection  <- table(snp_non_immune_selection_df$annotation_simplified)["Intron"]

## Construct a contingency table:
intron <- rbind(c(immune_intron_selection,
                  non_immune_intron_selection),
                c(total_selected_immune - immune_intron_selection,
                  total_selected_nonimmune - non_immune_intron_selection))
## Perform test:
intron_chi <- chisq.test(intron)
## Print result:
intron_chi
## Print observed values:
intron_chi$observed
## Print expected values:
intron_chi$expected

## Perform the same for SNP 5kb up/downstream of SNPs: 
immune_stream_selection  <- table(snp_immune_selection_df$annotation_simplified)["5kb up/down-stream"]
non_immune_stream_selection  <- table(snp_non_immune_selection_df$annotation_simplified)["5kb up/down-stream"]

## Construct a contingency table:
stream_table <- rbind(c(immune_stream_selection,
                  non_immune_stream_selection),
                c(total_selected_immune - immune_stream_selection,
                  total_selected_nonimmune - non_immune_stream_selection))
## Perform test:
stream_table_chi <- chisq.test(stream_table)
## Print result:
stream_table_chi
## Print observed values:
stream_table_chi$observed
## Print expected values:
stream_table_chi$expected

## Perform the same for SNPs in UTRs: 
immune_utr_selection  <- table(snp_immune_selection_df$annotation_simplified)["UTR"]
non_immune_utr_selection  <- table(snp_non_immune_selection_df$annotation_simplified)["UTR"]

## Construct a contingency table:
utr_table <- rbind(c(immune_utr_selection,
                  non_immune_utr_selection),
                c(total_selected_immune - immune_utr_selection,
                  total_selected_nonimmune - non_immune_utr_selection))
## Perform test:
utr_table_chi <- chisq.test(utr_table)
## Print result:
utr_table_chi
## Print observed values:
utr_table_chi$observed
## Print expected values:
utr_table_chi$expected

## Perform the same for SNPs in coding regions: 
immune_cds_selection  <- table(snp_immune_selection_df$annotation_simplified)["Coding"]
non_immune_cds_selection  <- table(snp_non_immune_selection_df$annotation_simplified)["Coding"]

## Construct a contingency table:
cds_table <- rbind(c(immune_cds_selection,
                     non_immune_cds_selection),
                   c(total_selected_immune - immune_cds_selection,
                     total_selected_nonimmune - non_immune_cds_selection))
## Perform test:
cds_table_chi <- chisq.test(cds_table)
## Print result:
cds_table_chi
## Print observed values:
cds_table_chi$observed
## Print expected values:
cds_table_chi$expected
```

10. To determine whether canonical immune genes are identified as under selection
in other datasets, we can load in the nSL scores for the population genomic 
analysis from the Colgan et al. (2022, MBE) - available in Supplemental Table S2.

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
## Read in data and store in an object:
british_nsl_data <- read.table(file = "data/selection_analyses/british_nsl_scores.txt",
                               header = FALSE)
## Update column names:
colnames(british_nsl_data) <- c("locus",
                                "nsl_score")

## Add the nSL scores from the British study to our dataset:
pop_data$british_nsl_score <- british_nsl_data[match(pop_data$locus,
                                                     british_nsl_data$locus), ]$nsl_score
pop_data$british_nsl_score <- gsub(pattern = ",",
                                   replacement = ".",
                                   pop_data$british_nsl_score)

## Ensure nSL scores are numeric:
pop_data$british_nsl_score <- as.numeric(as.character(unlist(pop_data$british_nsl_score)))

## Subset immune genes:
pop_data_immune <- subset(x = pop_data,
                          locus %in% immune_genes) 

## Generate plot:
nsl_comp_plot <- ggplot(data = pop_data_immune,
                        aes(x = nsl,
                            y = british_nsl_score)) +
  geom_point(alpha = 0) +
  geom_point(data = pop_data_immune,
             colour = "black",
             alpha = 0.5) +
  xlim(c(0, 4.5)) +
  ylim(c(0, 4.5)) +
  geom_point(data = subset(pop_data_immune,
                           nsl >= 2 &
                           british_nsl_score >= 2),
             colour = "blue",
             alpha = 0.5) +
  geom_point(data = subset(pop_data_immune,
                           nsl >= 2.51 &
                           british_nsl_score >= 2.56),
             colour = "red",
             alpha = 0.75) +
  geom_smooth(method = "lm",
              colour = "black",
              linetype = "dashed",
              alpha = 0.2) +
  geom_vline(xintercept = 2,
             linetype = "dashed",
             alpha = 0.5) +
  geom_hline(yintercept = 2,
             linetype = "dashed",
             alpha = 0.5) +
  labs(x = expression(bold("|")~bolditalic(n)*bold(S[L])*bold("| score - population 1"))) +
  labs(y = expression(bold("|")~bolditalic(n)*bold(S[L])*bold("| score - population 2"))) +
  stat_cor(p.accuracy = 0.0001,
           r.accuracy = 0.01) +
  theme_bw() +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold",
                                  size = 12))

## Generate a combined plot:
ggarrange(nsl_scatterplot,
          nsl_comp_plot,
          ncol = 2,
          nrow = 1,
          labels = c("A",
                     "B"),
          widths = c(1.5, 1),
          align = "hv")

## Save image:
ggsave(filename = "results/combined_plot_figure_1.pdf",
       height = 6,
       width = 10,
       dpi= 600)
```

11. Check style with lintr:

```{r, echo = FALSE, message = FALSE, results = 'hide', warning=FALSE}
require(lintr)
lintr::lint(file = "plot_nsl_scores_immune.Rmd")
```
