#!/bin/sh

#SBATCH --time=24:00:00
#SBATCH --nodes=1
#SBATCH -A mulif005c
#SBATCH -p ProdQ

#############################################################################################
## Author: Sarah Larragy, Joe Colgan (joscolgan)         	Program: run_freebayes.sh
##
## Date: 10/02/22
##
## Program:
## This script takes as input indexed bam files generated by bowtie2 and uses the variant
## caller freebayes to identify variants (e.g., polymorphic sites, indels) in population data.
## The script outputs a variant call format (VCF) file containing the physical locations
## of variants in the reference genome assembly, as well as the variant type, and genotypes
## called across individual samples.
##
#############################################################################################

## Load conda and activate conda environment:
module load conda/2
source activate conda_env

## Load parallel:
module load parallel

## Assign the relative path of the reference genome assembly to a variable: 
ref="data/Bombus_terrestris.Bter_1.0.dna.toplevel.fa"

## Run freebayes:
src/freebayes/scripts/freebayes-parallel \
region_list.txt 40 \
-f "$ref" \
--bam-list bam_list.txt \
--ploidy 2 \
--report-genotype-likelihood-max \
--use-mapping-quality \
--genotype-qualities \
--use-best-n-alleles 4 \
--haplotype-length 0 \
--min-base-quality 3 \
--min-mapping-quality 1 \
--min-alternate-frac 0.25 \
--min-coverage 1 \
--use-reference-allele > results/raw_samples.vcf
