#!/usr/bin/env Rscript
################################################################################################
##
## Author: Joe Colgan (joscolgan).                          Program: run_zscore_analysis.R
##
## Date: 20-04-2022
## 
## Purpose:
## The purpose of this script is to identify genomic regions of significantly reduced nucleotie diversity,
## which may be indicative of recent hard selective sweeps. The script takes as input a tab-delimited text file
## containing nucleotide diversity (pi) and Tajima's D estimates for 100kb sliding windows generated by 
## [PopGenome](https://cran.r-project.org/web/packages/PopGenome/index.html).  
## The script z-score transforms pop gen estimates for each windows and calculates a p-value per window.  
## The script outputs a table containing these p-values with each row representative of a 100kb sliding window.  
## The script also saves and outputs the entire analysis as an R object.
################################################################################################

## Take arguments from the command line:
args = commandArgs(trailingOnly=TRUE)

## Assign 'chromosome_info.txt' to input:
input <- args[1]
print(input)

## Assign third argument to output
output <- args[2]
print(output)

## Read in input file:
data <- read.table(file = input,
                   header = FALSE)
## Print first line of input dataframe:
print(head(data))
## Check structure:
print(str(data))

## Ensure colnames are correct:
colnames(data) <- c("chrom",
                    "start",
                    "end",
                    "nuc_diversity",
                    "tajima_d",
                    "seg_sites")

## Calculate z-score for nucleotide diversity:
data_nd_sd <- sd(data$nuc_diversity) *
                 sqrt((length(data$nuc_diversity) - 1) /
                              (length(data$nuc_diversity)))
print(head(data_nd_sd))
## Calculate mean nucleotide diversity:
data_nd_mean <- mean(data$nuc_diversity)

## Centre and scale:
data$nd_zscores <- (data$nuc_diversity - data_nd_mean) / data_nd_sd

## Calculate p value:
data$nd_pvalue2sided <- 2 * pnorm(-abs(data$nd_zscores))
## Calculate z-score for Tajima's D:
data_td_sd <- sd(data$tajima_d) *
              sqrt((length(data$tajima_d) - 1) /
                           (length(data$tajima_d)))
## Calculate mean Tajima's D:
data_td_mean <- mean(data$tajima_d)

## Centre and scale:
data$td_zscores <- (data$tajima_d - data_td_mean) / data_td_sd

## Calculate p value:
data$td_pvalue2sided <- 2 * pnorm(-abs(data$td_zscores))

## Write the output to file:

## Turn off scientific notation:
options(scipen = 999)

## write estimates for all windows to output:
write.table(x = data,
            file = output,
            row.names = FALSE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")

# Subset regions with significantly high or low estimates of :
# 1) nucleotide diversity;
# 2) Tajima's D.

## Subset significant terms:
data_nd_sig <- subset(data,
                      nd_pvalue2sided < 0.05)
write.table(x = data_nd_sig,
            file = paste(output,
                        "_all_sig.txt",
                        sep = ""),
            row.names = FALSE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")

## Subset significant terms with high zscores:
data_nd_high_sig <- subset(data,
                      nd_pvalue2sided < 0.05 &
                      nd_zscores > 0)
write.table(x = data_nd_high_sig,
            file = paste(output,
                        "_all_sig_high.txt",
                        sep = ""),
            row.names = FALSE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")

## Subset significant terms with low zscores:
data_nd_low_sig <- subset(data,
                      nd_pvalue2sided < 0.05 &
                      nd_zscores < 0)
write.table(x = data_nd_low_sig,
            file = paste(output,
                        "_all_sig_low.txt",
                        sep = ""),
            row.names = FALSE,
            col.names = TRUE,
            quote = FALSE,
            sep = "\t")
